Sources/SourceKitLSP/Swift/ReferenceDocument.swift

```diff
@@ -0,0 +1,150 @@
+package var supportedReferenceDocumentTypes: [ReferenceDocumentType: any ReferenceDocument.Type] = [
+  .macroExpansion: MacroExpansionReferenceDocument.self
+]
+
+/// Represents the document type of the `ReferenceDocument`
+package enum ReferenceDocumentType: String {
+  case macroExpansion = "macro-expansion"
+}
+
+/// Represents the parameter-value pairs of the reference document's uri
+package protocol ReferenceDocumentData {
+  var queryItems: [URLQueryItem]? { get }
+
+  init(from queryItems: [URLQueryItem]?) throws
+}
+
+package extension ReferenceDocument {
+  var url: URL {
+    get throws {
+      let baseURL = "sourcekit-lsp://\(Self.documentType.rawValue)"
+      guard var components = URLComponents(string: baseURL) else {
+        throw ReferenceDocumentError(
+          description: "Unable to create URL for ReferenceDocument: \(baseURL)"
+        )
+      }
+
+      components.path = "/\(displayName)"
+      components.queryItems = data.queryItems
+
+      guard let url = components.url else {
+        throw ReferenceDocumentError(
+          description: "Unable to create URL for ReferenceDocument"
+        )
+      }
+
+      return url
+    }
+  }
+
+  init(from url: URL) throws {
+    guard let components = URLComponents(string: url.absoluteString) else {
+      throw ReferenceDocumentError(
+        description: "Unable to parse the given url's components: \(url.absoluteString)"
+      )
+    }
+
+    guard url.scheme == "sourcekit-lsp" else {
+      throw ReferenceDocumentError(description: "Invalid Scheme for ReferenceDocument")
+    }
+
+    guard url.host(percentEncoded: false) == Self.documentType.rawValue else {
+      throw ReferenceDocumentError(description: "Invalid Document Type")
+    }
+
+    let data = try Data(from: components.queryItems)
+
+    self.init(data: data)
+  }
+
+  init(from urlString: String) throws {
+    guard let url = URL(string: urlString) else {
+      throw ReferenceDocumentError(description: "Unable to parse the given url: \(urlString)")
+    }
+
+    try self.init(from: url)
+  }
+}
+
+extension DocumentURI {
+  /// If `self` is a `DocumentURI` of a reference document, this returns the corresponding `ReferenceDocument`.
+  func referenceDocument() throws -> any ReferenceDocument {
+    guard let url = URL(string: self.stringValue) else {
+      throw ReferenceDocumentError(description: "Invalid URI for ReferenceDocument")
+    }
+
+    guard url.scheme == "sourcekit-lsp" else {
+      throw ReferenceDocumentError(description: "Invalid Scheme for ReferenceDocument")
+    }
+
+    guard let documentType = url.host(percentEncoded: false),
+      let referenceDocumentType = ReferenceDocumentType(rawValue: documentType),
+      let type = supportedReferenceDocumentTypes[referenceDocumentType]
+    else {
+      throw ReferenceDocumentError(description: "Invalid Document Type")
+    }
+
+    return try type.init(from: url)
+  }
+}
```

Comment:
With my proposal of `ReferenceDocumentURL`, this should be handled by the initializer.